src/Features/Core/Portable/ConvertTupleToStruct/AbstractConvertTupleToStructCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,865 @@
+                scopes.ToImmutableAndFree(),
+                isInlinable: false));
+        }
+
+        private CodeAction CreateAction(CodeRefactoringContext context, Scope scope)
+            => new MyCodeAction(GetTitle(scope), c => ConvertToStructAsync(context.Document, context.Span, scope, c));
+
+        private string GetTitle(Scope scope)
+        {
+            switch (scope)
+            {
+                case Scope.ContainingMember: return FeaturesResources.and_update_usages_in_containing_member;
+                case Scope.ContainingType: return FeaturesResources.and_update_usages_in_containing_type;
+                case Scope.ContainingProject: return FeaturesResources.and_update_usages_in_containing_project;
+                case Scope.DependentProjects: return FeaturesResources.and_update_usages_in_dependent_projects;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(scope);
+            }
+        }
+
+        private async Task<(SyntaxNode, INamedTypeSymbol)> TryGetTupleInfoAsync(
+            Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var position = span.Start;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            // Span actually has to be within the token (i.e. not in trivia around it).
+            if (!token.Span.IntersectsWith(position))
+            {
+                return default;
+            }
+
+            if (!span.IsEmpty && span != token.Span)
+            {
+                // if there is a selection, it has to be of the whole token.
+                return default;
+            }
+
+            var tupleExprNode = token.Parent as TTupleExpressionSyntax;
+            var tupleTypeNode = token.Parent as TTupleTypeSyntax;
+            if (tupleExprNode == null && tupleTypeNode == null)
+            {
+                return default;
+            }
+
+            var expressionOrType = tupleExprNode ?? (SyntaxNode)tupleTypeNode;
+
+            // The position/selection must be of the open paren for the tuple, or the entire tuple.
+            if (expressionOrType.GetFirstToken() != token)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var tupleType = semanticModel.GetTypeInfo(expressionOrType, cancellationToken).Type as INamedTypeSymbol;
+            if (tupleType?.IsTupleType != true)
+            {
+                return default;
+            }
+
+            return (expressionOrType, tupleType);
+        }
+
+        private async Task<Solution> ConvertToStructAsync(
+            Document document, TextSpan span, Scope scope, CancellationToken cancellationToken)
+        {
+            var (tupleExprOrTypeNode, tupleType) = await TryGetTupleInfoAsync(
```

Comment:
Nit: probably doesn't make a big difference, but you could avoid re-computing the tuple info (just pass it in).

Reply:
We dont' like doing that in code actions.  Because that can end up holding a compilation alive.  So the analysis to determine if there is a code aciton, and the code action operation itself should not share that same object.